BenchPress/Helpers/BenchPress.Azure/Authentication.psm1

```diff
@@ -0,0 +1,162 @@
+<#
+  .SYNOPSIS
+    Get-RequiredEnvironmentVariable is a private helper method that retrieves environment variables with the
+    expectation that if they are not present that an error will be logged with an immedate exit.
+
+  .DESCRIPTION
+    Get-RequiredEnvironmentVariable retrieves the environment variable specified by the input parameter and checks to
+    be sure that a value is present for that environment variable. If the value is missing or whitespace a message will
+    be written to Write-Error with the name of the variable in the output and exit will be called.
+
+  .PARAMETER VariableName
+    This is the name of the enviroment variable to retrieve and validate that a value is present.
+
+  .EXAMPLE
+    Provide -VariableName Parameter
+
+    Get-RequiredEnvironmentVariable -VariableName AZ_APPLICATION_ID
+
+  .EXAMPLE
+    Provide variable name without -VariableName Parameter
+
+    Get-RequiredEnvironmentVariable AZ_APPLICATION_ID
+
+  .INPUTS
+    System.String
+
+  .OUTPUTS
+    System.String
+#>
+function Get-RequiredEnvironmentVariable {
+  [OutputType([System.String])]
+  param (
+    [Parameter(Mandatory=$true, Position=0)]
+    [string]$VariableName
+  )
+  Begin {
+    $Value = [string]$null
+  }
+  Process {
+    $Value = [System.Environment]::GetEnvironmentVariable($VariableName)
+
+    if ([string]::IsNullOrWhiteSpace($Value)) {
+      Write-Error("Missing Required Environment Variable $VariableName")
+      exit 1
+    }
+  }
+  End {
+    return $Value
+  }
+}
+
+<#
+  .SYNOPSIS
+    Connect-Account uses environment variable values to log into an Azure context. This is an internal function and
+    should not be used outside of the module.
+
+  .DESCRIPTION
+    Connect-Account is designed to login to an Azure context using environment variables to login as a
+    ServicePrincipal for the PowerShell session.
+
+    The expected environment variables are:
+
+    AZ_APPLICATION_ID - The Service Principal ID
+    AZ_ENCRYPTED_PASSWORD - The Service Principal account password properly encrypted using ConvertTo-SecureString and
+                            saved as an environment variable using ConvertFrom-SecureString
+    AZ_TENANT_ID - The Tenant ID to login to
+    AZ_SUBSCRIPTION_ID - The Subscription ID to login to
+
+    If the current context that is logged in to matches the Service Principal, Tenant, and Subscription this function
+    is a no-op.
+
+  .EXAMPLE
+    There is only one way to call Connect-Account:
+
+    Connect-Account
+
+  .INPUTS
+    None
+
+  .OUTPUTS
+    None
+#>
+function Connect-Account {
```

Comment:
Since this is an internal function that is not intended to be used by the users, how will this function be called? Or is there another issue that will cover wiring this up, so that it is always called when users try to execute any of the other BenchPress functions?

Reply:
I was envisioning this being called within the modules, likely in the `Begin` block, for any cmdlets that will access Azure resources. There is likely a better approach for cross-cutting concerns like this in PowerShell, but this was the first approach that I thought of.